### ISSUE TITLE
reconnection problem when broker is restarted

### ISSUE THREAD

**Issue by heinsaris** [smartrent/tortoise311#18]
I have written an app that uses Tortoise to connect to a mosquitto broker(v2.0.14). Everything works fine until I stop the broker. At first I used publish directly from the place where the mqtt messages are created. Each time the broker is stopped, and then restarted, the messages in the inflight are sent. But then the client (Tortoise) closes the connection again and never recovers.

I get these errors in the console. First (only once)

15:56:31.059 [error] GenStateMachine {Tortoise311.Registry, {Tortoise311.Connection.Inflight, "HeinSim"}} terminating
** (exit) {:protocol_violation, :unknown_identifier}
    (stdlib 3.16.1) gen_statem.erl:1388: :gen_statem.loop_state_callback_result/11
    (stdlib 3.16.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
State: {{:connected, {Tortoise311.Transport.Tcp, #Port<0.14>}}, %Tortoise311.Connection.Inflight{client_id: "HeinSim", order: [], pending: %{}}}
Callback mode: :handle_event_function
Postponed events: []
Last event: {:cast, {:update, {:received, %Tortoise311.Package.Puback{__META__: %Tortoise311.Package.Meta{flags: 0, opcode: 4}, identifier: 27484}}}}
Queued events: []
State: {{:connected, {Tortoise311.Transport.Tcp, #Port<0.14>}}, %Tortoise311.Connection.Inflight{client_id: "HeinSim", order: [], pending: %{}}}

and then ( for each message in inflight)

15:56:31.416 [error] GenStateMachine {Tortoise311.Registry, {Tortoise311.Connection.Inflight, "HeinSim"}} terminating
** (CaseClauseError) no case clause matching: {:error, :closed}
    (tortoise311 0.11.1) lib/tortoise311/connection/inflight.ex:292: Tortoise311.Connection.Inflight.handle_event/4
    (stdlib 3.16.1) gen_statem.erl:1196: :gen_statem.loop_state_callback/11
    (stdlib 3.16.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
State: {{:connected, {Tortoise311.Transport.Tcp, #Port<0.14>}}, %Tortoise311.Connection.Inflight{client_id: "HeinSim", order: [17311], pending: %{17311 => %Tortoise311.Connection.Inflight.Track{caller: {#PID<0.316.0>, #Reference<0.3953562195.1766850567.237007>}, identifier: 17311, pending: [[dispatch: %Tortoise311.Package.Publish{__META__: %Tortoise311.Package.Meta{flags: 0, opcode: 3}, dup: false, identifier: 17311, payload: "....", qos: 1, retain: false, topic: "BH/1/670820"}, received: %Tortoise311.Package.Puback{__META__: %Tortoise311.Package.Meta{flags: 0, opcode: 4}, identifier: 17311}], [{:respond, {#PID<0.316.0>, #Reference<0.3953562195.1766850567.237007>}}, :cleanup]], polarity: :negative, status: [], type: Tortoise311.Package.Publish}}}}
Callback mode: :handle_event_function
Postponed events: []

Since the connection is setup again after the broker restarts, some messages go through, but then the connection is closed almost immediately:

1639148191: New client connected from 127.0.0.1:56500 as HeinSim (p2, c0, k60, u'u1').
1639148191: Client HeinSim closed its connection.

And it never recovers.

I then changed the implementation to add a queue and process the messages one by one via publish_sync, but with the same result. After the broker comes up again, the connections is restored but almost immediately closed, and never recovers.

The Tortoise connection is supervised by my app supervisor.

I am sure that I do something wrong, but I have no idea what. Is it because of the client-id? Hopefully someone can help me.

**Comment by jfcloutier** [comment]
Thanks for the issue report.

Could you please see if branch https://github.com/smartrent/tortoise311/tree/reconnect_fix takes care of your problem? If it doesn't, please provide me with the error log.

You might want to have a look at https://github.com/smartrent/jackalope. Jackalope simplifies the use of Tortoise311.

**Comment by heinsaris** [comment]
Hi,

Thank you for the fast response.

I tried your fix, and it works!

Does this mean that this was a bug in Tortoise? Or am I doing something wrong that caused this behaviour? 

Kind regards,

Hein Saris

On 10 December 2021 at 21:42:02, Jean-Francois Cloutier ***@***.***) wrote:

Thanks for the issue report.

Could you please see if branch https://github.com/smartrent/tortoise311/tree/reconnect_fix takes care of your problem? If it doesn't, please provide me with the error log.

You might want to have a look at https://github.com/smartrent/jackalope. Jackalope simplifies the use of Tortoise311.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by jfcloutier** [comment]
You are very welcome! 

The problem was with Tortoise311. It could crash if the MQTT broker sent unexpected data. It could also crash if the connection socket closed right after opening and Tortoise311 tried to use it before the disconnection was discovered.
